### PR TITLE
Workaround for zerg supply rounding bug

### PR DESCRIPTION
**Pull request by SoupCatcher**
See https://github.com/Blizzard/s2client-proto/issues/123

### CODE REVIEW

File: sc2/bot_ai.py [hunk 1]
```diff
@@ -137,6 +137,29 @@ def expansion_locations(self) -> Dict[Point2, Units]:
         """ Returns dict with the correct expansion position Point2 key, resources (mineral field, vespene geyser) as value """
         return centers
 
+    @property_cache_once_per_frame
+    def _supply_zerg_correction(self):
+        """ The client incorrectly rounds zerg supply down instead of up (see
```
[the remark this answers is not in the document]
Comment (SoupCatcher) [reply sc2/bot_ai.py]: Reported in November so I won't hold my breath...

File: sc2/bot_ai.py [hunk 2]
```diff
@@ -137,6 +137,19 @@ def expansion_locations(self) -> Dict[Point2, Units]:
...
+            https://github.com/Blizzard/s2client-proto/issues/123), so self.supply_used
+            and friends return the wrong value when there are an odd number of zerglings
+            and banelings. This function corrects the bad values. """
+        # TODO: remove when Blizzard/sc2client-proto#123 gets fixed.
+        correction = self.units({UnitTypeId.ZERGLING, UnitTypeId.ZERGLINGBURROWED,
+                                 UnitTypeId.BANELING, UnitTypeId.BANELINGBURROWED,
+                                 UnitTypeId.BANELINGCOCOON}).amount % 2
+        self.supply_used += correction
+        self.supply_army += correction
+        self.supply_left -= correction
```
Comment (tweakimp) [inline sc2/bot_ai.py]: You dont need to do `self.supply_left -= correction if you call the function before `self.supply_left` in` _prepare_step`

Comment (SoupCatcher) [reply sc2/bot_ai.py]: Yes, but that breaks up the nice grouping of the `supply_*` properties, making the code a little less readable. I'm not sure which I prefer.

Comment (tweakimp) [reply sc2/bot_ai.py]: yeah ok... :)

File: sc2/bot_ai.py [hunk 3]
```diff
@@ -588,6 +601,11 @@ def _prepare_step(self, state):
         self.supply_cap: Union[float, int] = state.common.food_cap
         self.supply_used: Union[float, int] = state.common.food_used
         self.supply_left: Union[float, int] = self.supply_cap - self.supply_used
+
+        # Workaround Zerg supply rounding bug
+        if self.race == Race.Zerg:
+            self._correct_zerg_supply()
```
Comment (tweakimp) [inline sc2/bot_ai.py]: Maybe be super smart and also move `self.larva_count: int = state.common.larva_count` here? :)

Comment (SoupCatcher) [reply sc2/bot_ai.py]: That feels like a separate PR to me...

Comment (tweakimp) [reply sc2/bot_ai.py]: That is the difference between you and me :p

**Pull request merged**